int-set/src/bitset.rs

```diff
@@ -0,0 +1,402 @@
+        self.page_map
+            .binary_search_by(|probe| probe.major_value.cmp(&major_value))
+            .ok()
+            .and_then(|info_idx| {
+                let real_idx = self.page_map[info_idx].index as usize;
+                self.pages.get(real_idx)
+            })
+    }
+
+    /// Return a mutable reference to the that 'value' resides in. Insert a new
+    /// page if it doesn't exist.
+    fn page_for_mut(&mut self, value: u32) -> &mut BitPage {
+        let major_value = self.get_major_value(value);
+        self.page_for_major_mut(major_value)
+    }
+
+    // Return a mutable reference to the page with major value equal to major_value.
+    // Inserts a new page if it doesn't exist.
+    fn page_for_major_mut(&mut self, major_value: u32) -> &mut BitPage {
+        match self
+            .page_map
+            .binary_search_by(|probe| probe.major_value.cmp(&major_value))
+        {
+            Ok(idx) => self
+                .pages
+                .get_mut(self.page_map[idx].index as usize)
+                .unwrap(),
+            Err(idx_to_insert) => {
+                let index = self.pages.len() as u32;
+                self.pages.push(BitPage::new_zeroes());
+                let new_info = PageInfo { index, major_value };
+                self.page_map.insert(idx_to_insert, new_info);
+                self.pages.last_mut().unwrap()
+            }
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+struct PageInfo {
+    // index into pages vector of this page
+    index: u32,
+    /// the top 23 bits of values covered by this page
+    major_value: u32,
+}
+
+impl<T> Hash for BitSet<T> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.iter_non_empty_pages().for_each(|t| t.hash(state));
+    }
+}
+
+impl<T> std::cmp::PartialEq for BitSet<T> {
+    fn eq(&self, other: &Self) -> bool {
+        let mut this = self.iter_non_empty_pages();
+        let mut other = other.iter_non_empty_pages();
+
+        // Note: normally we would prefer to use zip, but we also
+        //       need to check that both iters have the same length.
+        loop {
+            match dbg!(this.next(), other.next()) {
+                (Some(a), Some(b)) if a == b => continue,
+                (None, None) => return true,
+                _ => return false,
+            }
+        }
+    }
+}
+
+impl<T> std::cmp::Eq for BitSet<T> {}
+
+impl std::cmp::Ord for PageInfo {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.major_value.cmp(&other.major_value)
+    }
+}
+
+impl std::cmp::PartialOrd for PageInfo {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::collections::HashSet;
+
+    impl<T> BitSet<T> {
+        /// Return true if there are no members in this set.
+        pub fn is_empty(&self) -> bool {
+            self.len() == 0
+        }
+    }
```

Comment:
for less churn I would just keep this method in the main impl, but you can annotate it with either `#[allow(dead_code)]` or `#[cfg(test)]` there.